cla_backend/apps/reports/tests/test_mi_sla_report.py

```diff
@@ -29,43 +35,252 @@ def patch_field(cls, field_name, dt):
-                    "sla_15": get_sla_time(_dt, 15),
-                    "sla_30": get_sla_time(_dt, 30),
-                    "sla_120": get_sla_time(_dt, 120),
-                    "sla_480": get_sla_time(_dt, 480),
+                    "requires_action_at": requires_action_act,
+                    "sla_15": get_sla_time(requires_action_act, 15),
+                    "sla_30": get_sla_time(requires_action_act, 30),
+                    "sla_120": get_sla_time(requires_action_act, 120),
+                    "sla_480": get_sla_time(requires_action_act, 480),
                 },
             )
+            case.set_requires_action_at(requires_action_act)
 
-        case.requires_action_at = _dt
-        case.save()
-
+    def start_call(self, case, user, dt):
         event = event_registry.get_event("case")()
-        with patch_field(Log, "created", _make_datetime(2015, 1, 2, 9, 30, 0)):
+        with patch_field(Log, "created", dt):
             event.process(case, status="call_started", created_by=user, notes="Call started")
 
-        date_range = (_make_datetime(2015, 1, 1), _make_datetime(2015, 2, 1))
-
+    def get_report(self, date_range):
         with mock.patch("reports.forms.MICB1Extract.date_range", date_range):
             report = MICB1Extract()
 
             qs = report.get_queryset()
+            headers = report.get_headers()
+
+        return {k: v for k, v in zip(headers, qs[0])}
+
+    def create_and_get_report(self, callback_minutes_after):
+        dt = _make_datetime(2015, 1, 2, 9, 1, 0)
+        case = self.make_case(dt)
+        user = make_user()
+        make_recipe("call_centre.operator", user=user)
+
+        self.schedule_callback(case, user, dt)
+        self.start_call(case, user, dt + datetime.timedelta(minutes=callback_minutes_after))
+
+        date_range = (_make_datetime(2015, 1, 1), _make_datetime(2015, 2, 1))
+        return self.get_report(date_range)
+
+
+class MiSlaTestCaseWeb(MiSlaTestCaseBase):
+    source = CASE_SOURCE.WEB
+
+    def test_call_answered_within_sla1_window(self):
+        values = self.create_and_get_report(25)
+        self.assertFalse(values["missed_sla_1"])
+        self.assertFalse(values["missed_sla_2"])
+
+    def test_call_answered_after_sla1_window(self):
+        values = self.create_and_get_report(35)
+        self.assertTrue(values["missed_sla_1"])
+        self.assertFalse(values["missed_sla_2"])
+
+    def test_call_answered_before_sla1_window(self):
+        values = self.create_and_get_report(-5)
+        self.assertTrue(values["missed_sla_1"])
+        self.assertFalse(values["missed_sla_2"])
+
+    def test_call_answered_after_sla2_window(self):
+        values = self.create_and_get_report(75 * 60)
+        self.assertTrue(values["missed_sla_1"])
+        self.assertTrue(values["missed_sla_2"])
+
+    def test_cb1_uncalled_and_current_time_before_sla1(self):
+        # SLA 1 miss is FALSE - No action before sla1 window
```

Comment:
I don't understand `No action before sla1 window`

Reply:
I was trying to convey that the current time is before the SLA1 window and nothing has been done in contacting the user. i.e Callback is scheduled for 9:30-10 and the current time is 9AM.

I have now changed this to read `# SLA 1 miss is FALSE - Current time before SLA1 window`